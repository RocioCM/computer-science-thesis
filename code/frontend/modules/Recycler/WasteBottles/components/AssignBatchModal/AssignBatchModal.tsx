import Button from '@/common/components/Button';
import Modal, { ModalProps } from '@/common/components/Modal';
import { toast } from 'react-toastify';
import useForm from '@/common/hooks/useForm';
import { ASSIGN_FORM_INPUTS, ASSIGN_FORM_STRUCT } from '../../constants';
import WasteBottlesServices from '../../services';
import { AssignBottleToBatchPayload } from '../../types';
import { useEffect, useState } from 'react';
import { ZERO_ADDRESS } from '@/common/constants';
import { Option } from '@/common/components/Inputs/InputDropdown/InputDropdown';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const AssignBatchModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  const [availableBatchesList, setAvailableBatchesList] = useState<Option[]>(
    []
  );
  const { form, submitEnabled, formBuilder } = useForm(ASSIGN_FORM_STRUCT);

  const fetchAvailableBatches = async () => {
    const { ok, data } = await WasteBottlesServices.getAllUserRecyclingBatches(
      1,
      100
    ); /// TODO: implement infinite scroll or search.
    if (ok) {
      const filteredData = data
        ?.filter(
          (batch) => !batch.buyerOwner || batch.buyerOwner === ZERO_ADDRESS
        )
        .map((batch) => ({
          label: `Lote #${batch.id}`,
          value: batch.id,
        }));
      setAvailableBatchesList(filteredData ?? []);
    } else {
      toast.error('Ocurrió un error al cargar los lotes disponibles');
    }
  };

  useEffect(() => {
    fetchAvailableBatches();
  }, []);

  const handleSubmit = async () => {
    const payload: AssignBottleToBatchPayload = {
      bottleId: editingId as number,
      batchId: form.batchId,
    };
    const { ok } = await WasteBottlesServices.assignBottleToBatch(payload);
    if (ok) {
      toast.success(
        `Botella asignada al lote #${payload.batchId} correctamente`
      );
      handleCancel();
      handleSuccess();
    } else {
      toast.error(
        'Ocurrió un error al asignar la botella. Por favor, inténtelo de nuevo más tarde'
      );
    }
  };

  return (
    <Modal handleCancel={handleCancel} {...props}>
      <h2 className="w-full font-medium text-[1.75rem] bg-n0 border-b border-n2 pb-2 mb-1">
        Asignar lote #{editingId}
      </h2>
      <div className="w-full flex flex-col gap-m my-xl">
        {formBuilder(ASSIGN_FORM_INPUTS, {
          dropdownOptions: { batchId: availableBatchesList },
        })}
      </div>
      <Button
        handleClick={handleSubmit}
        disabled={!submitEnabled}
        width="full"
        label="Asignar"
      />
    </Modal>
  );
};

export default AssignBatchModal;
