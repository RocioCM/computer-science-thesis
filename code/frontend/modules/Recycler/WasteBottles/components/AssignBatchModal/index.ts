import AssignBatchModal from './AssignBatchModal';

export default AssignBatchModal;
