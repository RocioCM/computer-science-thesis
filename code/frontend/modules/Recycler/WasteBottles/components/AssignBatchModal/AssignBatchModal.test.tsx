import { render } from '@testing-library/react';
import AssignBatchModal from './AssignBatchModal';

describe('Recycler Waste Bottles - AssignBatchModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Matches snapshot', () => {
    const { container } = render(
      <div id="modal-root">
        <AssignBatchModal
          editingId={1}
          handleCancel={() => {}}
          handleSuccess={() => {}}
        />
      </div>
    );
    expect(container).toMatchSnapshot();
  });
});
