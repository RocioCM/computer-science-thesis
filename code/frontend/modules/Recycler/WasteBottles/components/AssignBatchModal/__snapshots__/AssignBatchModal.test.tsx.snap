// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recycler Waste Bottles - AssignBatchModal Matches snapshot 1`] = `
<div>
  <div
    id="modal-root"
  >
    <div
      class="bg-n10 bg-opacity-70 py-2xl px-xl w-full h-screen fixed top-0 left-0 z-modal flex justify-center items-center animate__animated animate__fadeIn cursor-pointer"
      data-testid="modal"
    >
      <div
        class="bg-n0 min-w-[300px] min-h-[200px] rounded-rm overflow-y-auto flex flex-col items-center relative cursor-default animate__animated animate__zoomIn w-[380px] h-max max-h-full max-w-full px-3xl py-3xl"
      >
        <div
          class="flex justify-end w-full h-0 sticky top-0 z-10"
        >
          <img
            alt="Cerrar"
            class="h-6 w-auto cursor-pointer object-contain"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
        </div>
        <h2
          class="w-full bg-n0"
        >
          Asignar lote #
          1
        </h2>
        <div
          class="w-full flex flex-col gap-m my-xl"
        >
          <div
            class="w-full"
          >
            <span
              class="block relative z-10 w-max transition-all text-p mb-xs cursor-pointer top-0 left-0 text-n10 font-semibold"
            >
              ID del lote
            </span>
            <button
              class="relative flex flex-row flex-nowrap justify-between items-center w-full border bg-transparent outline-none transition-shadow gap-xs p text-left rounded-rs border-n3 px-m py-m placeholder:text-n2 h-10 hover:border-n2 dropdonwFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2"
              data-testid="dropdown"
              name="batchId"
            >
              <span
                class="flex-1 min-h-4"
              >
                Selecciona un lote
              </span>
              <i
                class="inline-block w-2.5 h-2.5 rotate-0 "
                data-testid="chevron-icon"
              >
                <svg
                  class="w-full h-full block transition-transform duration-300 ease-out"
                  fill="none"
                  viewBox="0 0 15 9"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M14.7 1.89992L13.5 0.919922L7.50001 6.09992L1.50001 0.919921L0.300013 1.89992L7.50001 8.19992L14.7 1.89992Z"
                    fill="var(--colors-neutrals-n10)"
                  />
                </svg>
              </i>
            </button>
          </div>
        </div>
        <button
          class="relative flex items-center justify-center gap-s rounded-rs font-semibold h-10 disabled:cursor-default disabled:bg-opacity-40  focusButton bg-p1 text-n0 py-s px-m w-full"
          data-testid="button"
          disabled=""
        >
          <span>
            Asignar
          </span>
        </button>
      </div>
      <style>
        
          body {
            overflow: hidden;
          }
        
      </style>
    </div>
  </div>
</div>
`;
